dlt/destinations/impl/sqlalchemy/db_api_client.py

```diff
@@ -0,0 +1,378 @@
+from dlt.destinations.impl.sqlalchemy.configuration import SqlalchemyCredentials
+from dlt.common.typing import TFun
+
+
+class SqlaTransactionWrapper(DBTransaction):
+    def __init__(self, sqla_transaction: sa.engine.Transaction) -> None:
+        self.sqla_transaction = sqla_transaction
+
+    def commit_transaction(self) -> None:
+        if self.sqla_transaction.is_active:
+            self.sqla_transaction.commit()
+
+    def rollback_transaction(self) -> None:
+        if self.sqla_transaction.is_active:
+            self.sqla_transaction.rollback()
+
+
+def raise_database_error(f: TFun) -> TFun:
+    @wraps(f)
+    def _wrap_gen(self: "SqlalchemyClient", *args: Any, **kwargs: Any) -> Any:
+        try:
+            return (yield from f(self, *args, **kwargs))
+        except Exception as e:
+            raise self._make_database_exception(e) from e
+
+    @wraps(f)
+    def _wrap(self: "SqlalchemyClient", *args: Any, **kwargs: Any) -> Any:
+        try:
+            return f(self, *args, **kwargs)
+        except Exception as e:
+            raise self._make_database_exception(e) from e
+
+    if inspect.isgeneratorfunction(f):
+        return _wrap_gen  # type: ignore[return-value]
+    return _wrap  # type: ignore[return-value]
+
+
+class SqlaDbApiCursor(DBApiCursorImpl):
+    def __init__(self, curr: sa.engine.CursorResult) -> None:
+        # Sqlalchemy CursorResult is *mostly* compatible with DB-API cursor
+        self.native_cursor = curr  # type: ignore[assignment]
+        curr.columns
+
+        self.fetchall = curr.fetchall  # type: ignore[method-assign]
+        self.fetchone = curr.fetchone  # type: ignore[method-assign]
+        self.fetchmany = curr.fetchmany  # type: ignore[method-assign]
+
+    def _get_columns(self) -> List[str]:
+        return list(self.native_cursor.keys())  # type: ignore[attr-defined]
+
+    # @property
+    # def description(self) -> Any:
+    #     # Get the underlying driver's cursor description, this is mostly used in tests
+    #     return self.native_cursor.cursor.description  # type: ignore[attr-defined]
+
+    def execute(self, query: AnyStr, *args: Any, **kwargs: Any) -> None:
+        raise NotImplementedError("execute not implemented")
+
+
+class DbApiProps:
+    # Only needed for some tests
+    paramstyle = "named"
+
+
+class SqlalchemyClient(SqlClientBase[Connection]):
+    external_engine: bool = False
+    dialect: sa.engine.interfaces.Dialect
+    dialect_name: str
+    dbapi = DbApiProps  # type: ignore[assignment]
+
+    def __init__(
+        self,
+        dataset_name: str,
+        staging_dataset_name: str,
+        credentials: SqlalchemyCredentials,
+        capabilities: DestinationCapabilitiesContext,
+        engine_args: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        super().__init__(credentials.database, dataset_name, staging_dataset_name, capabilities)
+        self.credentials = credentials
+
+        if credentials.engine:
```

Comment:
we could also reuse a single engine. you could look at DuckDbConnection which creates a single master connection and then borrows instances to clients.

hmmm we do not need to overcomplicate. I hope opening an engine is not that heavy

Reply:
Yeah I was trying to keep it simple for now, but definitely would be nice to use the connection pool.   
I had in mind some generic solution for this cause it applies to other destinations too. E.g. if we add a "thread safe" capability then the loader can re-use the whole jobclient instance across jobs. In this case it wouldn't take much to make the client thread friendly, just a matter of moving the "current transaction" stuff to context vars.